src/bundle/bundle.ts

```diff
@@ -24,54 +33,78 @@ export interface IBundleProps {
-  const { bundleConfig, ctx } = props;
+export function createBundle(props: IBundleProps): Bundle {
+  const { bundleConfig, ctx, priority, type, webIndexed = true } = props;
   const outputConfig = ctx.outputConfig;
   const isProduction = ctx.config.isProduction;
   const target = ctx.config.target;
 
-  const bundleWriter = distWriter({ hashEnabled: isProduction, root: outputConfig.root });
+  const bundleWriter = distWriter({ hashEnabled: isProduction, root: outputConfig.distRoot });
   const apiCore = props.includeAPI && {
     interopRequireDefault: ctx.compilerOptions.esModuleInterop,
     isIsolated: bundleConfig.isolatedApi,
     target,
   };
-  const source = bundleSource({ core: apiCore, isIsolated: bundleConfig.isolatedApi, target });
+  const source = createBundleSource({ core: apiCore, isIsolated: bundleConfig.isolatedApi, target });
   source.entries = props.entries;
 
-  const self: IBundle = {
+  const self: Bundle = {
+    config: null,
+    contents: null,
+    data: null,
+    priority,
     source,
-    type: props.type,
+    type,
+    webIndexed,
+    addCodeSplittingMap: (map: ICodeSplittingMap) => {
+      self.source.codeSplittingMap = map;
+    },
+    createSourceMap: async () => {
+      const sourceMapName = path.basename(self.config.relativePath) + '.map';
+      self.contents += `\n//# sourceMappingURL=${sourceMapName}`;
+      const targetDir = path.dirname(self.config.absPath);
+      const sourceMapFile = path.join(targetDir, sourceMapName);
+      await bundleWriter.write(sourceMapFile, self.data.sourceMap);
+    },
     generate: async () => {
-      const data = source.generate();
-
-      const options = { contents: data.content.toString(), hash: isProduction, userString: bundleConfig.path };
-      const config = bundleWriter.createWriter(options);
-
-      let contents = options.contents;
-      if (source.containsMaps) {
-        // writing source maps
-        const sourceMapName = path.basename(config.relativePath) + '.map';
-        contents = options.contents + `\n//# sourceMappingURL=${sourceMapName}`;
-        const targetDir = path.dirname(config.absPath);
-        const sourceMapFile = path.join(targetDir, sourceMapName);
-        await bundleWriter.write(sourceMapFile, data.sourceMap);
+      if (!self.config) {
+        self.prepare();
       }
-
-      await bundleWriter.write(config.absPath, contents);
-
-      const bundleResponse = { absPath: config.absPath, bundle: self, relativePath: config.relativePath };
-      return [bundleResponse];
+      // writing source maps
+      if (source.containsMaps) self.createSourceMap();
+      // write the bundle to fs
+      await bundleWriter.write(self.config.absPath, self.contents);
+      return [
+        {
+          absPath: self.config.absPath,
+          // bundle: self,
+          relativePath: self.config.relativePath,
+        },
+      ];
+    },
+    prepare: (): IWriterConfig => {
+      self.data = source.generate();
```

Comment:
we shouldn't actually generate the content at this stage. IT should be generate with the API (which isn't ready yet)
So technically we just prepare the paths